src/compiler/checker.ts

```diff
@@ -7555,6 +7559,62 @@ namespace ts {
             return undefined;
         }
 
+        function shouldAcquireLiteralType(literalNode: StringLiteral) {
+
+            let current: Node = literalNode;
+            while (true) {
+                const { parent } = current;
+                switch (parent.kind) {
+                    // The operand of a 'switch' should get a literal type.
+                    case SyntaxKind.SwitchStatement:
+                        return current === (parent as SwitchStatement).expression;
+
+                    // The tested expression of a 'case' clause should get a literal type.
+                    case SyntaxKind.CaseClause:
+                        return current === (parent as CaseClause).expression;
+
+                    case SyntaxKind.BinaryExpression:
+                        const binaryExpr = parent as BinaryExpression;
+                        switch (binaryExpr.operatorToken.kind) {
+                            // Either operand of an equality/inequality comparison
+                            // should get a literal type.
+                            case SyntaxKind.EqualsEqualsEqualsToken:
+                            case SyntaxKind.ExclamationEqualsEqualsToken:
+                            case SyntaxKind.EqualsEqualsToken:
+                            case SyntaxKind.ExclamationEqualsToken:
+                                return current === binaryExpr.left || current === binaryExpr.right;
+
+                            case SyntaxKind.BarBarToken:
+                                current = parent;
+                                continue;
+
+                        }
+                        break;
+
+                    case SyntaxKind.ConditionalExpression:
+                    case SyntaxKind.ParenthesizedExpression:
+                        current = parent;
+                        continue;
+                }
+
+                // This is not a node we can account for.
+                // Let contextual typing take over.
+                break;
+            }
+
+            // We haven't found a "literal match location" (i.e. a location that signals
+            // a literal should get a literal type). Check whether the contextual type
+            // has a literal type in it.
+            //
+            // We could perform our walk, check if 'current' is an expression when we get to a
+            // a node that isn't a match location, and then get the contextual type of that.
+            // That would save a few steps but the checks in 'isExpression' seem so involved
+            // that it would probably be better to simply grab the contextual type if we didn't.
+            const contextualType = getContextualType(literalNode);

```

Comment:
I guess I assumed too much when I wrote the above comment. `getContextualType` expects an `Expression`, so we'd need to check if `current` is an expression using `isExpression`. But that function is so roundabout, I figured we'd be better off just performing the walk from `literalNode` again.

Additionally, you need to because we just skipped `||`, where the RHS gets contextually typed by the LHS if a parent contextual type doesn't exist.